packages/compass-components/src/hooks/use-toast.tsx

```diff
@@ -0,0 +1,149 @@
+import React, { createContext, useCallback, useContext, useState } from 'react';
+import type { ToastVariant } from '..';
+import { css } from '..';
+import { Toast } from '..';
+
+type ToastProperties = {
+  title?: React.ReactNode;
+  body: React.ReactNode;
+  variant: ToastVariant;
+  progress?: number;
+  timeout?: number;
+};
+
+type ToastState = ToastProperties & {
+  timeoutRef?: ReturnType<typeof setTimeout>;
+};
+
+interface ToastActions {
+  openToast: (id: string, toastProperties: ToastProperties) => void;
+  closeToast: (id: string) => void;
+}
+
+const ToastContext = createContext<ToastActions>({
+  openToast: () => {
+    console.log('Fake openToast');
+
+    //
+  },
+  closeToast: () => {
+    console.log('Fake closeToast');
+    //
+  },
+});
+
+const toastStyles = css({
+  button: {
+    position: 'absolute',
+  },
+});
+
+/**
+ * @example
+ *
+ * ```
+ * const MyButton = () => {
+ *   const { openToast } = useToast('namespace');
+ *   return <button onClick={() => openToast(
+ *      'myToast1', {title: 'This is a notification'})} />
+ * };
+ *
+ * <ToastArea><MyButton/><ToastArea>
+ * ```
+ *
+ * @returns
+ */
+export const ToastArea: React.FunctionComponent = ({ children }) => {
+  const [toasts, setToasts] = useState<Record<string, ToastState>>({});
+
+  const closeToast = useCallback(
+    (toastId: string): void => {
+      console.log('real closeToast');
+
+      const { timeoutRef } = toasts[toastId] || {};
+      if (timeoutRef) {
+        clearTimeout(timeoutRef);
+      }
+
+      setToasts((prevToasts) => {
+        const newToasts = { ...prevToasts };
+        delete newToasts[toastId];
+        return newToasts;
+      });
+    },
+    [toasts, setToasts]
+  );
+
+  const openToast = useCallback(
+    (toastId: string, toastProperties: ToastProperties): void => {
+      console.log('real openToast');
+
+      // if updating clear timeouts first
+      const { timeoutRef } = toasts[toastId] || {};
+      if (timeoutRef) {
+        clearTimeout(timeoutRef);
+      }
+
+      setToasts((prevToasts) => ({
+        ...prevToasts,
+        [toastId]: {
+          ...toastProperties,
+          timeoutRef: toastProperties.timeout
+            ? setTimeout(() => {
+                closeToast(toastId);
```

Comment:
is this ok? do we need to do something to cancel this timeout on unmount?

Reply:
Oh, yeah, that's actually a good point. You do want to clean up the timeouts on unmount. I think if you want to go full galaxy brain / idiomatic hooks usage on this, you would probably even want this timeout management be out of this state and done by something separately, something with an interface like this:

```js
const timeoutManager = useTimeoutManager((id) => {
  // This will be some sort of onTimeout callback
  closeToast(id);
});

const openToast = useCallback(() => {
  ...
  // This will keep a reference to the timeout in some sort of internal timeout manager map or something
  timeoutManager.startTimeout(id, timeout)
})
```

And then the code inside `useTimeoutManager` can do the clean-up on unmount. This will make a robust, single purpose hook that's easily reusable somewhere else.

... or tbh you can just have something like this

```js
useEffect(() => {
  return () => {
    Object.values(toasts).forEach(({timeoutRef}) => timeoutRef && clearTimeout(timeoutRef))
  }
}, [])
```

and call it a day. Eslint will scream at you because the dependencies list is not matching, but you would want it like that so that the clean-up callback is only called on unmount and not every time that toasts state is changing

Reply:
updated this one, you may want to take a look :)